content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
+well add one.
+
+We have a working compiler to armv8 with no register allocation, and are working
+on linear scan register allocation in arm as well as x86 code
+generation. Unfortunately, these features aren't quite finished due to being
+undergraduate students with other finals, however there's a good chance they'll
+be fixed after the class is over (in the interest of having something nice to
+merge into the bril repo).
+
+Our assembly does depend on the C standard library for functions like `malloc`
+(to support the memory extension) as well as a few printing functions as doing
+it all with system calls just didn't seem like a great use of time. We have some
+functions in the `lib/` directory for printing base types, as well as parsing
+them in order for the `main` function to read command line arguments. A future
+improvement would be putting `malloc` and other C functions we call into this
+file so that we could keep a compiled version around and not have to link with
+C.
+
+## Implementation
+Our interpreter consists of an initial pass to translate into this bytecode,
+which turns temps and labels from strings into numbers, and reduces the size of
+the code overall. The bytecode instructions are stored in an array so that we
+can benefit from cache locality. We then interpret them by iterating through
+this array. One minor inconvenience is that certain instructions, such as print,
+require multiple words. This means that our array index for iteration will
+NOT be an induction variable, so the C compiler will not be able to optimize it
+as aggressively. One solution to this could be using a less flat representation 
+of a list of instructions, and using pointers for the multiword 
+instructions, but this seems like more trouble than it's worth, especially 
+considering the benefits of cache locality.
+
+For the armv8 assembly generation, we initially implemented it as a single pass
+over the bytecode that printed the instructions as it went. This is fine, but it
+leaves little room for extensions and improvements, so in order to get started
+on register allocation we refactored to have an internal representation of arm
+instructions. For this, we went with a tagged union, as it is similar to a
+variant which is very convenient for this, as learned in CS 4120.
+
+## Evaluation
+We used the Bril benchmarks from the repo, and used clock time to evaluate the
+difference. This isn't exactly optimal as most of these benchmarks are very
+small, so the startup time might be significant, but the startup time should
+hurt our interpreter the most, as it has the translation phase. Our results are
+[here](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/eval/performance.ods),
+in ods format. On average, `brili-rs` has an 84% speedup over `brili`, and
+`fastbrili` has a 77% speedup over `brili-rs`. `fastbrili` also has a 96%
+speedup over `brili`, so I think we can call this a job well done.
```

Comment:
I don't think you mean a *speedup* of 77% = 0.77, which would indicate that fastbrili was slower. ["Speedup"](https://en.wikipedia.org/wiki/Speedup) is defined as `<baseline time>/<your time>`, so it is greater than 1 when your thing is faster.

Also, what do you mean by "average"? Remember to use the geometric mean or harmonic mean, never the arithmetic mean, when averaging ratios…

Reply:
ok i was doing something stupid where i did (fast - slow)/slow. no idea where i got this. fastbril runs in .02 of the time of brili, and .23 of the time of brilirs